### PR TITLE
Implements profiling for GD2

### PR DESCRIPTION
**Pull request by harigowtham**
Currently CPU and memory/heap profiling are implemented.
Changes have been made to enabled them from the config file.
The output prof file is seen is /tmp/profile...

The output file has to be passed to the go tool pprof to read and
understand the bottleneck we are trying to resolve.

Signed-off-by: Hari Gowtham <hgowtham@redhat.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
@harigowtham CI failure PTAL

**Comment by harigowtham** [comment]
@Madhu-1 The profiling package is not available on the CI machines. They have to be installed on them.

**Comment by harigowtham** [comment]
addresses issue: https://github.com/gluster/glusterd2/issues/1344

**Comment by nigelbabu** [comment]
retest this please

**Comment by nigelbabu** [comment]
retest this please

**Comment by nigelbabu** [comment]
retest this please

**Comment by atinmu** [comment]
@harigowtham 

CI failure

11:58:24 ./glusterd2/utils/pprof.go:12:11: should omit type string from declaration of var basePath; it will be inferred from the right-hand side
11:58:24 ./glusterd2/utils/pprof.go:18:1: exported function EnableProfiling should have comment or be unexported
11:58:24 Found 2 lint suggestions; failing.
11:58:32 failed ./test/020-golint.sh [1]

File: glusterd2/utils/pprof.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+	"github.com/gorilla/mux"
+)
+
+var (
+	basePath string = "/debug/pprof/"
+)
+
+func addPath(router *mux.Router, name string, handler http.Handler) {
+	router.Path(basePath + name).Name(name).Handler(handler)
```
Comment (Madhu-1) [inline glusterd2/utils/pprof.go]: add a debug log to know what are the registered profiling handlers.

File: glusterd2/utils/pprof.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+func EnableProfiling(router *mux.Router) {
```
Comment (Madhu-1) [inline glusterd2/utils/pprof.go]: add function comment as this is an exported function

File: glusterd2/utils/pprof.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"net/http"
+	"net/http/pprof"
+	runtime_pprof "runtime/pprof"
+
+	"github.com/gorilla/mux"
+)
+
+var (
+	basePath string = "/debug/pprof/"
```
Comment (Madhu-1) [inline glusterd2/utils/pprof.go]: no need to explicitly mention it as a string 
`basePath  = "/debug/pprof/"` should work

File: glusterd2/utils/pprof.go (hunk 1 above)
Comment (harigowtham) [reply glusterd2/utils/pprof.go]: done

File: glusterd2/utils/pprof.go (hunk 2 above)
Comment (harigowtham) [reply glusterd2/utils/pprof.go]: done

File: glusterd2/utils/pprof.go (hunk 3 above)
Comment (harigowtham) [reply glusterd2/utils/pprof.go]: done

**Comment by Madhu-1** [comment]
@harigowtham need to update commit message.

**Comment by harigowtham** [comment]
@Madhu-1 Have fixed the typo in commit message.

File: glusterd2/servers/rest/rest.go [hunk 4]
```diff
@@ -85,6 +86,12 @@ func NewMuxed(m cmux.CMux) *GDRest {
 
 	rest.registerRoutes()
 
+	//Enable go profiling
+	profiling := config.GetBool("profiling")
+	if profiling == true {
```
Comment (Madhu-1) [inline glusterd2/servers/rest/rest.go]:
```
if profiling {
	gdutils.EnableProfiling(rest.Routes)
}
```

File: glusterd2/utils/pprof.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+var (
+	basePath = "/debug/pprof/"
+)
+
+func addPath(router *mux.Router, name string, handler http.Handler) {
+	router.Path(basePath + name).Name(name).Handler(handler)
+	log.WithFields(log.Fields{
+		"name":    name,
+		"handler": handler,
+	}).Debug("Starting profile for %s", name)
```
Comment (Madhu-1) [inline glusterd2/utils/pprof.go]: the name is logged twice in this log message.
once can be removed

File: glusterd2/servers/rest/rest.go (hunk 4 above)
Comment (harigowtham) [reply glusterd2/servers/rest/rest.go]: done

File: glusterd2/utils/pprof.go (hunk 5 above)
Comment (harigowtham) [reply glusterd2/utils/pprof.go]: done

File: glusterd2/utils/pprof.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	basePath = "/debug/pprof/"
+)
+
+func addPath(router *mux.Router, name string, handler http.Handler) {
+	router.Path(basePath + name).Name(name).Handler(handler)
+	log.WithField("handler", handler).Debug("Starting profile for %s", name)
```
Comment (Madhu-1) [inline glusterd2/utils/pprof.go]: @harigowtham can you please check how this log gets printed, I think you need to print basePath+Name

Comment (aravindavk) [reply glusterd2/utils/pprof.go]: Any variable should be part of log fields. Log message should be constant

**Comment by aravindavk** [comment]
PR Looks good, please update developer nodes about using this feature

File: glusterd2/utils/pprof.go (hunk 6 above)
Comment (harigowtham) [reply glusterd2/utils/pprof.go]: Have made the debug log message constant

> PR Looks good, please update developer nodes about using this feature

Do you want the developer note's update as a part of this patch or a different one?

Comment (Madhu-1) [reply glusterd2/utils/pprof.go]: it good to have it with this PR. https://github.com/heketi/heketi/pull/1396 this will help you to document.

Comment (harigowtham) [reply glusterd2/utils/pprof.go]: its along with the example, will create a new file as mentioned and update it

**Pull request merged**